### DISCUSSION TITLE
avro schema namespace/name vs apicurio artifact groupId/name

### DISCUSSION THREAD

**Discussion by jangalinski** [Apicurio/apicurio-registry#2309] [Q&A]
Is it safe to assume that when my avro schema uses namespace="foo" and name="Bar" that it is a good idea to use groupId="foo" and name="Bar" as well when I upload the schema file?
If so ... couldn't the registry automatically derive namespace/name from an avro schema, so we do not risk typos when registering?

**Comment by EricWittmann** [comment]
The answer to the first is "yes" in most cases.  I think it typically makes sense, but not always.  You can use the "group" concept in registry for whatever you want (e.g. group artifacts by team or by project), so it probably depends on the use-case.

The second part is trickier.  The API doesn't currently lend itself well to what you are suggesting.  Because you would be using a REST call like this:

`POST /apis/registry/v2/groups/foo/artifacts`

This would result in an artifact created in group "foo".  I think we could introduce a new API endpoint perhaps that allowed auto-discovery of both the artifactId and the groupId.  So it's an interesting idea.

@Apicurio/developers any thoughts?

**Comment by jsenko** [comment]
We need to consider how would we derive the group & artifact names from other schema types. Avro has the `namespace` and Protobuf has the `package` concepts, but JSON schema and OpenAPI do not. We would have to limit that endpoint to only accept some artifact types, or invent a non-standard way to specify group names, e.g. using `x-*` properties in case of OpenAPI schema.

**Reply by EricWittmann** [reply]
Yes good points.  Some types won't have this.  This is actually a tricky thing to think about.  I might argue that the qualified name should just be used as the artifactId.  So instead of groupId=`foo` and artifactId=`bar` we would just have artifactId=`foo.bar`.  This would require no changes to the API, just another option to the Create Artifact operation.  That operation **already** generates an artifactId if none is provided.  We could just have an option to extract the artifactId from the content (and perhaps fail or use a UUID if it can't).
